modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/io/Metadata.scala

```diff
@@ -0,0 +1,237 @@
+import com.snowplowanalytics.iglu.core.circe.implicits._
+import com.snowplowanalytics.snowplow.scalatracker.{Emitter, Tracker}
+import com.snowplowanalytics.snowplow.scalatracker.emitters.http4s.Http4sEmitter
+import com.snowplowanalytics.snowplow.enrich.common.fs2.config.io.{Metadata => MetadataConfig}
+import com.snowplowanalytics.snowplow.enrich.common.outputs.EnrichedEvent
+
+trait Metadata[F[_]] {
+  def report: Stream[F, Unit]
+  def observe(event: EnrichedEvent): F[Unit]
+}
+
+object Metadata {
+
+  private implicit def unsafeLogger[F[_]: Sync]: Logger[F] =
+    Slf4jLogger.getLogger[F]
+
+  def build[F[_]: ContextShift: ConcurrentEffect: Timer](
+    config: MetadataConfig,
+    reporter: MetadataReporter[F]
+  ): F[Metadata[F]] =
+    MetadataEventsRef.init[F].map { observedRef =>
+      new Metadata[F] {
+        def report: Stream[F, Unit] =
+          for {
+            _ <- Stream.eval(Logger[F].info("Starting metadata repoter"))
+            _ <- Stream.fixedDelay[F](config.interval)
+            snapshot <- Stream.eval(MetadataEventsRef.snapshot(observedRef))
+            _ <- Stream
+                   .emits[F, MetadataEvent](snapshot.mappings.keySet.toSeq)
+                   .evalMap[F, Unit](reporter.report(snapshot.periodStart, snapshot.periodEnd)(_, snapshot.mappings))
+          } yield ()
+
+        def observe(event: EnrichedEvent): F[Unit] =
+          observedRef.mappings.update(recalculate(_, event))
+      }
+    }
+
+  def noop[F[_]: Async]: Metadata[F] =
+    new Metadata[F] {
+      def report: Stream[F, Unit] = Stream.never[F]
+      def observe(event: EnrichedEvent): F[Unit] = Applicative[F].unit
+    }
+
+  trait MetadataReporter[F[_]] {
+    def report(periodStart: Instant, periodEnd: Instant)(snapshot: MetadataEvent, mappings: Map[MetadataEvent, Set[SchemaKey]]): F[Unit]
+  }
+
+  case class HttpMetadataReporter[F[_]: ConcurrentEffect: Timer](
+    config: MetadataConfig,
+    appName: String,
+    client: Client[F]
+  ) extends MetadataReporter[F] {
+    def initTracker(
+      config: MetadataConfig,
+      appName: String,
+      client: Client[F]
+    ): Resource[F, Tracker[F]] =
+      for {
+        emitter <- Http4sEmitter.build(
+                     Emitter.EndpointParams(
+                       config.endpoint.host.map(_.toString()).getOrElse("localhost"),
+                       config.endpoint.port,
+                       https = config.endpoint.scheme.map(_ == Uri.Scheme.https).getOrElse(false)
+                     ),
+                     client,
+                     retryPolicy = Emitter.RetryPolicy.MaxAttempts(10),
+                     callback = Some(emitterCallback _)
+                   )
+      } yield new Tracker(NonEmptyList.of(emitter), "tracker-metadata", appName)
+
+    def report(periodStart: Instant, periodEnd: Instant)(event: MetadataEvent, mappings: Map[MetadataEvent, Set[SchemaKey]]): F[Unit] =
+      initTracker(config, appName, client).use { t =>
```

Comment:
Here you initialize a new tracker and emitter every time you want to report.  That's fine -- I don't think there's much overhead in initializing multiple times.  An alternative implementation would have been to initialize once, and use the same emitter for every report.